pkg/utils/test/matchers/managedresource.go

```diff
@@ -20,36 +20,65 @@ import (
 	"strings"
 
 	"github.com/onsi/gomega/format"
+	"golang.org/x/exp/maps"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 
 	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
 )
 
-type managedResourceDataMatcher struct {
-	ctx         context.Context
-	cl          client.Client
-	expectedObj client.Object
+type managedResourceObjectsMatcher struct {
+	ctx                     context.Context
+	cl                      client.Client
+	expectedObjToSerialized map[client.Object]string
+	extraObjectsCheck       bool
 
-	expectedObjectSerialized string
-	actualObjectSerialized   string
-	secretNames              []string
+	extraObjects             []string
+	missingObjects           []string
+	mismatchExpectedToActual map[string]string
 }
 
-func (m *managedResourceDataMatcher) FailureMessage(_ interface{}) string {
-	if m.actualObjectSerialized != "" {
-		return format.Message(m.actualObjectSerialized, "to equal", m.expectedObjectSerialized)
-	}
-	return format.Message(m.expectedObjectSerialized, fmt.Sprintf("to be found in managed resource secrets %v", m.secretNames))
+func (m *managedResourceObjectsMatcher) FailureMessage(actual interface{}) string {
+	return m.createMessage(actual, "not to be")
+}
+
+func (m *managedResourceObjectsMatcher) NegatedFailureMessage(actual interface{}) string {
+	return m.createMessage(actual, "to be")
 }
 
-func (m *managedResourceDataMatcher) NegatedFailureMessage(_ interface{}) string {
-	return format.Message(m.expectedObjectSerialized, fmt.Sprintf("not to be found in managed resource secrets %v", m.secretNames))
+func (m *managedResourceObjectsMatcher) createMessage(actual interface{}, addition string) string {
+	managedResource, ok := actual.(*resourcesv1alpha1.ManagedResource)
+	if !ok {
+		return fmt.Errorf("expected *resourcesv1alpha1.ManagedResource.  got:\n%s", format.Object(actual, 1)).Error()
```

Comment:
Why do you use `fmt.Errorf(...).Error()` instead of simply doing `fmt.Printf(...)`?